### ISSUE TITLE
@Delegate is not working for method with different generic overloads

### ISSUE THREAD

**Issue by tde-desc** [projectlombok/lombok#1442]
Given:
```
public interface AbstractEntityRepository {
    Entity findOne(Long id);

    void delete(Long id);

    Entity saveAndFlush(Entity entity);
}

public interface EventRepository extends JpaRepository<Entity, Long> {
    
}

public class EventRepositoryDelegee implements AbstractEntityRepository {
    @Delegate
    private final EventRepository repository;
}
```

- AbstractEntityRepository#saveAndFlush is being delegated correctly to EventRepository#saveAndFlush
- AbstractEntityRepository#findOne and AbstractEntityRepository#delete cannot be compiled, even though JpaRepository is extending CrudRepository which declares those two methods. 
org.apache.maven.plugins:maven-compiler-plugin:3.6.0:compile says: 

> EventRepositoryDelegee is not abstract and does not override abstract method delete(java.lang.Long) in AbstractEntityRepository.

Same compiler error for findOne.

I guess the problem is, that there are multiple overloads for findOne and delete with different generic arguments ...

Any idea?

**Comment by tde-desc** [comment]
Still being an issue ...

**Comment by omega09** [comment]
Can you post a code example which we can all (try to) compile?

**Comment by tde-desc** [comment]
[demo.zip](https://github.com/rzwitserloot/lombok/files/1345945/demo.zip)

Building the project with maven results in the following error:

com.example.lombok.overloading.demo.client.EventRepositoryDelegee is not abstract and does not override abstract method delete(java.lang.Long) in com.example.lombok.overloading.demo.third.party.AbstractEntityRepository

**Comment by omega09** [comment]
No problem on my end. I tested with a simple example:

```
interface Ext {
	Integer get(Integer i);
}

interface Comp<T> {
	T get(T t);
}

interface Base extends Comp<Integer> {}

class All implements Ext {

	@Delegate
	Base comp;
}

class Test {

	public static void main(String[] args) {
		All all = new All();
		System.out.println(all.get(0));
	}
}
```
Maven or Spring don't seem to matter. Try the above code.  I'm using Eclipse, java 8, and Lombok 1.16.18.

**Comment by tde-desc** [comment]
Your example is missing my point. Change Comp<T> as follows and you should face the same problem:

interface Comp<T, ID extends Serializable> {
	T get(T t);
        T get(ID id);
}

I guess the main difference regarding the behaviour is the fact, that Comp provides two overloads of "get" having the same signature after types have been erased.

If you're not able to reproduce it like that, pls take a closer look at my demo.zip above, which is having a similar setup and resulting in the mentioned problem.

**Comment by omega09** [comment]
If I use your `Comp` and use `Integer` as the replacement for `ID` then the error is `The method get(Integer) is ambiguous for the type Base` (as it should), which is different from what you have.  Otherwise there's no problem.

Your Maven example, as I said, gives no error for me and the error you are reporting for it is different than the error you are reporting for this one, which indicates that you're doing something different. You need to create a simple example like I did which produces the same error to be used for comparison.

Also have a look at the [Delegate](https://projectlombok.org/features/experimental/Delegate) page which talks about problems with generics.

**Comment by tde-desc** [comment]
I'm wondering why `mvn clean install` is not generating the mentioned error on your machine. So lets change my demo.zip in the following way:

- I removed the spring-data dependency
- I tried to reproduce the same interface-hierarchy as spring-data does

The new [demo2.zip](https://github.com/rzwitserloot/lombok/files/1347055/demo2.zip) is producing the mentioned error just as I pointed out:

`/demo/src/main/java/com/example/lombok/overloading/demo/client/EventRepositoryDelegee.java:[9,8] com.example.lombok.overloading.demo.client.EventRepositoryDelegee is not abstract and does not override abstract method findOne(java.lang.Long) in com.example.lombok.overloading.demo.third.party.two.ExpectedEntityRepository`

I'm afraid @Delegate is not able to analyze the interface-hierarchy of the third.party.one-package

**Comment by omega09** [comment]
I created a [CompactDemo.zip](https://github.com/rzwitserloot/lombok/files/1347359/CompactDemo.zip)
 based on your demo. I removed completely the Spring dependencies (you still had them) and minimized the amount of code needed to produce the error. Please verify that you're getting the same error after a clean install.

If you do, then you'll notice that the error comes from having a generic interface bridging the method, `interface EntityRepository<T>`. If you either remove the `<T>` or add the method `void findOne(Long id);` to `EntityRepository` you should get a successful compilation. This is a bug because the generic parameter is meaningless yet changes the behavior.

This is a duplicate of #1431 and possibly even #485.

The high level reason this happens with `@Delegate` is because it is handled at a later stage of the compilation and maven does not pick it up properly.

**Comment by tde-desc** [comment]
I downloaded your CompactDemo.zip and executed `mvn clean install` receiving the following error:

`/demo/src/main/java/com/example/lombok/overloading/demo/client/EventRepositoryDelegee.java:[9,8] com.example.lombok.overloading.demo.client.EventRepositoryDelegee is not abstract and does not override abstract method findOne(java.lang.Long) in com.example.lombok.overloading.demo.client.ExpectedEntityRepository
`

**Comment by amareensaleh** [comment]
I ran into the same problem today...Is this planned on being fixed any time soon?

**Comment by mageddo** [comment]
I made a workaround for this, afraid it will only work when you have only one generic parameter type

* [MyArrayListDelegate.java](https://github.com/mageddo/java-examples/blob/2cd40e5c2c4d7c5326c0c41be95485896af4e05e/lombok-examples/src/main/java/com/mageddo/lombok/vo/delegate/MyArrayListDelegate.java)
* [MyArrayListDelegateTest.java](https://github.com/mageddo/java-examples/blob/2cd40e5/lombok-examples/src/test/java/com/mageddo/lombok/vo/delegate/MyArrayListDelegateTest.java)

**Comment by renedewaele** [comment]
> I made a workaround for this, just put the generic on the method return instead of on the class

@mageddo This still gives me a `java: Compilation failed: internal java compiler error` in java 20.

**Comment by Rawi01** [comment]
This was fixed in #2459
